runtime/vm/FastJNI_java_lang_ref_Reference.cpp

```diff
@@ -51,12 +51,33 @@ Fast_java_lang_ref_Reference_reprocess(J9VMThread *currentThread, j9object_t rec
 	}
 }
 
+/* java.lang.ref.Reference: static private native boolean waitForReferenceProcessingImpl(); */
+jboolean JNICALL
+Fast_java_lang_ref_Reference_waitForReferenceProcessingImpl(J9VMThread *currentThread, j9object_t classObject)
+{
+#if defined(J9VM_GC_FINALIZATION)
+	J9JavaVM *javaVM = currentThread->javaVM;
+	J9MemoryManagerFunctions *mmFuncs = javaVM->memoryManagerFunctions;
+	if (0 == mmFuncs->j9gc_wait_for_reference_processing(javaVM)) {
+		return JNI_FALSE;
+	} else {
+		return JNI_TRUE;
+	}
+#else
+	return JNI_FALSE;
+#endif
+}
+
+
 J9_FAST_JNI_METHOD_TABLE(java_lang_ref_Reference)
 	J9_FAST_JNI_METHOD("getImpl", "()Ljava/lang/Object;", Fast_java_lang_ref_Reference_getImpl,
 		J9_FAST_JNI_RETAIN_VM_ACCESS | J9_FAST_JNI_NOT_GC_POINT | J9_FAST_JNI_NO_NATIVE_METHOD_FRAME | J9_FAST_JNI_NO_EXCEPTION_THROW |
 		J9_FAST_JNI_NO_SPECIAL_TEAR_DOWN | J9_FAST_JNI_DO_NOT_WRAP_OBJECTS)
 	J9_FAST_JNI_METHOD("reprocess", "()V", Fast_java_lang_ref_Reference_reprocess,
 		J9_FAST_JNI_RETAIN_VM_ACCESS | J9_FAST_JNI_NOT_GC_POINT | J9_FAST_JNI_NO_NATIVE_METHOD_FRAME | J9_FAST_JNI_NO_EXCEPTION_THROW |
 		J9_FAST_JNI_NO_SPECIAL_TEAR_DOWN | J9_FAST_JNI_DO_NOT_WRAP_OBJECTS)
+	J9_FAST_JNI_METHOD("waitForReferenceProcessingImpl", "()Z", Fast_java_lang_ref_Reference_waitForReferenceProcessingImpl,
+		J9_FAST_JNI_RETAIN_VM_ACCESS | J9_FAST_JNI_NOT_GC_POINT | J9_FAST_JNI_NO_NATIVE_METHOD_FRAME | J9_FAST_JNI_NO_EXCEPTION_THROW |
```

Comment:
Does this ever need to wait for a GC to complete?  If so, we can deadlock as this method doesn't give up vm access

Reply:
no, it only waits for processing one reference in slave finalization thread(outside GC cycle), if currently the thread is clearing the reference.